core/client/fs/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -117,49 +118,54 @@ public BlockInfo getInfo(long blockId) {
       throw new NotFoundException("Block " + blockId + " is not available in Alluxio");
     }
+    WorkerNetAddress address = null;
+    if (blockInfo.getLocations().isEmpty()) {
+      BlockLocationPolicy blockLocationPolicy = Preconditions
+          .checkNotNull(options.getUfsReadLocationPolicy(),
+              PreconditionMessage.UFS_READ_LOCATION_POLICY_UNSPECIFIED);
+      address = blockLocationPolicy.getWorker(
+          GetWorkerOptions.defaults().setBlockWorkerInfos(getWorkerInfoList()).setBlockId(blockId)
+              .setBlockSize(blockInfo.getLength()));
+    }
+
     // TODO(calvin): Get location via a policy.
     // Although blockInfo.locations are sorted by tier, we prefer reading from the local worker.
     // But when there is no local worker or there are no local blocks, we prefer the first
     // location in blockInfo.locations that is nearest to memory tier.
     // Assuming if there is no local worker, there are no local blocks in blockInfo.locations.
     // TODO(cc): Check mContext.hasLocalWorker before finding for a local block when the TODO
     // for hasLocalWorker is fixed.
-    if (Configuration.getBoolean(PropertyKey.USER_SHORT_CIRCUIT_ENABLED)) {
-      for (BlockLocation location : blockInfo.getLocations()) {
-        WorkerNetAddress workerNetAddress = location.getWorkerAddress();
-        if (workerNetAddress.getHost().equals(mLocalHostName)) {
-          // There is a local worker and the block is local.
-          try {
-            return StreamFactory.createLocalBlockInStream(mContext, blockId, blockInfo.getLength(),
-                workerNetAddress, options);
-          } catch (Exception e) {
-            LOG.warn("Failed to open local stream for block {}: {}", blockId, e.getMessage());
-            // Getting a local stream failed, do not try again
-            break;
-          }
-        }
+    for (BlockLocation location : blockInfo.getLocations()) {
+      WorkerNetAddress workerNetAddress = location.getWorkerAddress();
+      if (workerNetAddress.getHost().equals(mLocalHostName)) {
+        address = workerNetAddress;
+        break;
       }
     }
-    // No local worker/block, choose a random location. In the future we could change this to
-    // only randomize among locations in the highest tier, or have the master randomize the order.
-    List<BlockLocation> locations = blockInfo.getLocations();
-    WorkerNetAddress workerNetAddress =
-        locations.get(mRandom.nextInt(locations.size())).getWorkerAddress();
+    if (address == null) {
+      // No local worker/block, choose a random location. In the future we could change this to
+      // only randomize among locations in the highest tier, or have the master randomize the order.
+      List<BlockLocation> locations = blockInfo.getLocations();
+      address = locations.get(mRandom.nextInt(locations.size())).getWorkerAddress();
+    }
     return StreamFactory
```

Comment:
NIT: do we still need `StreamFactory` now? it becomes very simple with only two methods, and I think the logic can be consolidated here rather than having factory logic in different classes. Up to you